javascripts/discourse/initializers/post-height-detective.js

```diff
@@ -14,80 +14,61 @@ export default {
 
         init() {
           this._super(...arguments);
-          this.recordedHeights = {};
-          this.recordedHtml = {};
+          this.recordedHeights = new WeakMap();
+          this.recordedHtml = new WeakMap();
+          this.resizeObserver = new ResizeObserver(this.resizeObserverCallback);
         },
 
         afterRender() {
           // Do this super quick whenever rerendering happens,
           // otherwise we might miss something
-          this.checkForHeightChanges();
+          this.trackElements();
         },
 
-        checkForHeightChanges() {
+        trackElements() {
           this.element
             .querySelectorAll("article[data-post-id]")
-            .forEach(this.trackHeight);
+            .forEach((e) => this.resizeObserver.observe(e));
         },
 
-        tick() {
-          const nextTick = later(
-            this,
-            () => {
-              this.checkForHeightChanges();
-              this.tick();
-            },
-            1000
-          );
-          this.set("nextTick", nextTick);
-        },
-
-        didInsertElement() {
-          this._super(...arguments);
-          this.tick();
-        },
-
-        willDestroyElement() {
-          this._super(...arguments);
-          const nextTick = this.nextTick;
-          cancel(nextTick);
-        },
-
-        _refresh() {
-          this._super(...arguments);
-          this.recordedHeights = {};
-          this.recordedHtml = {};
+        @bind
+        resizeObserverCallback(resizeObserverEntries) {
+          for (const entry of resizeObserverEntries) {
+            this.trackHeight(entry.target, entry.contentRect.height);
+          }
         },
 
         @bind
-        trackHeight(post) {
-          const postId = post.id.split("_")[1];
+        trackHeight(post, renderedHeight) {
+          const postNumber = post.id.split("_")[1];
           const topicId = this.get("posts.posts.firstObject.topic.id");
-          const postLink = `${window.location.protocol}//${window.location.hostname}/t/${topicId}/${postId}`;
+          const postLink = `${window.location.protocol}//${window.location.hostname}/t/${topicId}/${postNumber}`;
           const postElement = post.querySelector(".cooked");
-          const renderedHeight = postElement.getBoundingClientRect().height;
-          const oldHeight = this.recordedHeights[postId];
+          const oldHeight = this.recordedHeights.get(post);
```

Comment:
Is there a reason why it is better to use the post element as key instead of the `postNumber`? I do not think it ever happens for the post element to be replaced with a new one (the current one is always updated), but it would be more resilient I assume if it used `postNumber` instead.

This question applies to the other maps too.

Reply:
The problem is that keying on `postNumber` is not unique across topics. When the loading slider is enabled, the `scrolling-post-stream` component is reused when navigating directly between topics.

So it would measure the height of `#post_1` in `/t/1`, then you navigate to `/t/2` and it would compare it to the height of `#post_1` in that topic. I think a lot of other stuff would break if the post element was replaced, so I think on balance mapping on the element is more reliable.